Add CI migration blog post (#10532)
facebookincubator/velox: website/blog/2024-07-25-ci-migration.mdx

---
slug: ci-migration
title: "Optimizing and Migrating Velox CI Workloads to Github Actions"
authors: [jwujciak, kgpai]
tags: [tech-blog, packaging]
---

<figure>
 <img src="/img/abstract-white-background.jpg" height= "100%" width="100%"/>

> </figure>

Reviewer: While a nice image, I am not sure it really adds anything?

Author: yep, that's correct. We can remove it as well. Please let me know.

Author: cc @kgpai for thoughts

## TL;DR

In late 2023, the Meta OSS (Open Source Software) Team requested all Meta teams to move the CI deployments from CircleCI to Github Actions. [Voltron Data](http://voltrondata.com) and Meta in collaboration migrated all the deployed Velox CI jobs. For the year 2024, Velox CI spend was on track to overshoot the allocated resources by a considerable amount of money. As part of this migration effort, the CI workloads were consolidated and optimized by Q2 2024, bringing down the projected 2024 CI spend by 51%.

## Velox’s Continuous Integration Workload

> Continuous Integration (CI) is crucial for Velox’s success as an open source project as it helps protect from bugs and errors, reduces likelihood of conflicts and leads to increased community trust in the project. This is to ensure the Velox builds works well on a myriad of system architectures, operating systems and compilers - along with from the ones used internally at Meta. The OSS build version of Velox also supports additional features that aren't used internally in Meta (for example, support for Cloud blob stores, etc.).

Reviewer: ```suggestion
Continuous Integration (CI) is crucial for Velox’s success as an open source project as it helps protect from bugs and errors, reduces likelihood of conflicts and leads to increased community trust in the project. This is to ensure the Velox builds works well on a myriad of system architectures, operating systems and compilers - along with the ones used internally at Meta. The OSS build version of Velox also supports additional features that aren't used internally in Meta (for example, support for Cloud blob stores, etc.).
```

Author: thanks, updated now

When a pull request is submitted to Velox, the following jobs are executed:

1. Linting and Formatting workflows:
    a. Header checks
    b. License checks

>     c. Basic Linters 

Reviewer: Ordered lists can only be numeric. I think an unordered list (-/*) is sufficient from a contents perspective as well.
![image](https://github.com/user-attachments/assets/6ecc19c0-fb47-4018-8685-6df336a5b869)

Author: thanks, updated

2. Ensure Velox builds on various platforms

>     a. MacOS (13/14, M1)

Reviewer: ```suggestion
    a. MacOS (Intel, M1)
```

Author: thanks, updated

    b. Linux (Ubuntu/Centos)
3. Ensure Velox builds under its various configurations
    a. Debug / Release builds
    b. Build default Velox build

>     c.Build Velox with support for Parquet/Arrow/ External Adapters (S3/HDFS/GCS etc.)

Reviewer: ```suggestion
    c. Build Velox with support for Parquet, Arrow and External Adapters (S3/HDFS/GCS etc.)
```

Author: thanks, updated

    d. PyVelox builds
4. Run prerequisite tests 
    a. Unit Tests
    b. Benchmarking Tests
        i. [Conbench](https://velox-conbench.voltrondata.run/runs/5bd139fffa9b4e0eb020da4d63211121/) is used to store and compare results, and also alert users on regressions
    c. Various Fuzzer Tests (Expression / Aggregation/ Exchange / Join etc) 
    d. Signature Check and Biased Fuzzer Tests ( Expression / Aggregation)
    e. Fuzzer Tests using Presto as source of truth
5. Docker Image build jobs
    a. If an underlying dependency is changed, a new Docker CI image is built for
        i. Ubuntu Linux
        ii. Centos
        iii. Presto Linux image
6. Documentation build and publish Job
    a. If underlying documentation is changed, Velox documentation pages are rebuilt and published
    b. Netlify is used for publishing Velox web pages

## Velox CI Optimization

> Previous implementation of CI in CircleCI grew organically and was unoptimized, resulting in long build times and also, significantly costlier. This opportunity to migrate to Github Actions helped to take a holistic view of CI deployments and actively optimized to reduce build times and CI spend. Note that however there has been continued investment in increasing test times to further increase Velox reliability and stability. Some of the optimizations completed are below:-  

Reviewer: ```suggestion
Previous implementation of CI in CircleCI grew organically and was unoptimized, resulting in long build times, and also significantly costlier. This opportunity to migrate to Github Actions helped to take a holistic view of CI deployments and actively optimized to reduce build times and CI spend. Note however, that there has been continued investment in reducing test times to further improve Velox reliability, stability and developer experience. Some of the optimizations completed are:  
```

Author: thanks, updated
